File: go/data_structure/2_chapter/bubbleSort.go
package main

import "fmt"

func main() {
	for true {
		fmt.Println("please input array:")
		var arr []int
		for true {
			var i int
			_, err := fmt.Scan(&i)
			if err != nil {
				break
			}
			arr = append(arr, i)
		}
		bubbleSort(arr, 0, len(arr))
		fmt.Println("the bubble sorted array:", arr)
	}
}

func bubbleSort(arr []int, lo, hi int) {
	for lo < hi {
		lo = bubbleMin(arr, lo, hi)
		hi = bubbleMax(arr, lo, hi)
	}
}

func bubbleMax(arr []int, lo, hi int) int {
	last := lo
	lo++
	for ; lo < hi; lo++ {
		if arr[lo-1] > arr[lo] {
			last = lo
			arr[lo-1], arr[lo] = arr[lo], arr[lo-1]
		}
	}
	return last
}

func bubbleMin(arr []int, lo, hi int) int {
	first := hi
	hi--
	for ; lo < hi; hi-- {
		if arr[hi] < arr[hi-1] {
			first = hi
			arr[hi], arr[hi-1] = arr[hi-1], arr[hi]
		}
	}
	return first
}

File: cpp/data_structures/two_chapter/bubbleSort.c
template <typename T>
void bubblesort(int lo, int hi) {
	while ((lo = bubbleMin(lo, hi)) < (hi = bubble(lo, hi)));
}

template <typename T>
int bubbleMax(int lo, int hi) {
	int last = lo;
	while (++lo < hi)
		if (_elem[lo - 1] > _elem[lo]) {
			last = lo;
			swap (_elem[lo - 1], _elem[lo]);
		}
	return last;
}
template <typename T>
int bubbleMin(int lo, int hi) {
	int firts = hi;
	while (lo < --hi)
		if (_elem[hi] < _elem[hi - 1]) {
			first = hi;
			swap (_elem[hi], _elem[hi - 1]);
		}
	return first;
}
